### PR TITLE
Clean up and document metrics [databricks]

### LINKED ISSUE #2527: [FEA] clean up timing metrics and make them consistent
For now I am marking this as a task, but it might turn into an epic.

The goal is to expose timing metrics to the end user that are simple to understand and useful.  Along with this we should have a good way to calculate these metrics so they don't become inconsistent or difficult to deal with.

The proposed metrics are

|*NAME*|*LEVEL*|*DESCRIPTION*|
|-----------|------------|----------------------|
|`gpuOpTime`|MODERATE|The amount of time spent with the GPU semaphore held, ideally doing some kind of computation, but not totally a requirement. This will include time spent spilling data `spillWriteTime` to make room, and time spend reading back in spilled data `spillReadTime`. This is not going to be simple to measure directly because with spilling the semaphore can be acquired when trying to read a batch. So this will be calculated from `opTime` and `semaphoreWaitTime`|
|`opTime`|DEBUG|The amount of time spent either waiting for the GPU semaphore or actually doing computation. It should not include any calls to `hasNext`, `next`, or time spent reading data/writing data as a part of shuffle or to get input to computation/write the result of computation.|
|`semaphopreWaitTime`|DEBUG|The amount of time spent by this part of the plan stuck waiting to acquire the GPU semaphore. |
|`spillWriteTime`|DEBUG|The amount of time spent spilling data to make room. This is going to be a little bit tricky because all operations can have this happen, so we almost have to register a global thread local call back so when this does happen we can capture it.|
|`spillReadTime`|DEBUG|The amount of time spent reading back in spilled data. **WE NEED TO DECIDE DO WE INCLUDE ALLOCATING THE DATA ON THE GPU OR NOT** because if we do then it could also overlap with `spillWriteTime`|
|`fetchTime`|DEBUG/MODERATE|Time spent fetching data from upstream. This could be reading data from a task that is feeding it or it could be reading data from a distributed file system/etc.|

Special care will need to be taken with `opTime` and `semaphoreWaitTime` because `gpuOpTime` is calculated from the other two, but has a different metric level from them, and the NoopMetric does not actually store any data, it just throws it all away, so we will have to come up with a good class to wrap the three metrics and do the right thing in all cases.

### LINKED ISSUE #3243: Remove total time metrics from most GPU nodes in the plan
Many exec nodes in the plan have a "total time" metric that records the entire time spent in that node, including time spent waiting for the input iterator(s) `hasNext` and `next` calls.  I'm not sure this is very useful in practice, and many users think this metric represents time spent solely operating on the node, when often it is more measuring the time spent waiting for inputs.  Almost all the exec nodes have a separate metric that records the time spent locally operating on the GPU which in practice is much more useful.

Given we've seen cases where too many metrics have been generated while executing the plan, we should consider removing this one across the board, or at least only keeping it for a select few exec nodes where it makes sense.

### PR DESCRIPTION
**Pull request by revans2**
This fixes #3243 
This fixes #2527

This removes total time and replaces it with `op time`, among other metrics. I am open to modifying/adding in other metrics like in #2527.

### CODE REVIEW

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
build

File: build/buildall [hunk 1]
```diff
@@ -39,7 +39,7 @@ SPARK_SHIM_VERSIONS="
 BUILD_PARALLELISM=1
 
 time (
-echo -n "$SPARK_SHIM_VERSIONS" | xargs -I% -P $BUILD_PARALLELISM -n 1 \
+echo -n "$SPARK_SHIM_VERSIONS" | xargs -t -I% -P $BUILD_PARALLELISM -n 1 \
```
Comment (abellina) [inline build/buildall:42]: nit, debug?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala [hunk 2]
```diff
@@ -56,32 +61,38 @@ abstract class AbstractGpuJoinIterator(
...
-      } else if (hasNextStreamBatch) {
-        // Need to refill the gatherer
-        gathererStore.foreach(_.close())
-        gathererStore = None
-        gathererStore = setupNextGatherer(startNanoTime)
-        nextCb = nextCbFromGatherer()
       } else {
-        mayContinue = false
+        // The only part we want to skip in the timing is hasNextStreamBatch
+        val tmp = timedHasNextStreamBatch
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala:78]: nit, would be nice to decompose the return of `timedHasNextStreamBatch` here.

File: build/buildall (hunk 1 above)
Comment (revans2) [reply build/buildall:42]: From just looking at the build log it is very hard to tell what version of Spark is being built against. This gives a very clear clue.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala [hunk 3]
```diff
@@ -371,28 +371,44 @@ case class GpuFilterExec(condition: Expression, child: SparkPlan)
 /**
  * Physical plan for range (generating a range of 64 bit numbers).
  */
-case class GpuRangeExec(range: org.apache.spark.sql.catalyst.plans.logical.Range,
+case class GpuRangeExec(
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:374]: this one seems different than other changes in this PR. Just curious on why the change.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuCartesianProductExec.scala [hunk 4]
```diff
@@ -71,7 +70,8 @@ class GpuSerializableBatch(batch: ColumnarBatch)
   }
 
   private def readObject(in: ObjectInputStream): Unit = {
-    GpuSemaphore.acquireIfNecessary(TaskContext.get())
+    // TODO I don't know of a way to make this work...
```
Comment (abellina) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuCartesianProductExec.scala:73]: needs follow up issue?

File: build/buildall (hunk 1 above)
Comment (abellina) [reply build/buildall:42]: ack, sounds good

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala [hunk 5]
```diff
@@ -36,8 +35,8 @@ import org.apache.spark.sql.vectorized.ColumnarBatch
 abstract class AbstractGpuJoinIterator(
     gatherNvtxName: String,
     targetSize: Long,
-    joinTime: GpuMetric,
-    totalTime: GpuMetric) extends Iterator[ColumnarBatch] with Arm with AutoCloseable {
+    opTime: GpuMetric,
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala:38]: the scaladoc needs updating in this constructor

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala [hunk 6]
```diff
@@ -144,14 +155,13 @@ abstract class SplittableJoinIterator(
     builtBatch: LazySpillableColumnarBatch,
     targetSize: Long,
     spillCallback: SpillCallback,
-    joinTime: GpuMetric,
-    streamTime: GpuMetric,
-    totalTime: GpuMetric)
+    opTime: GpuMetric,
+    joinTime: GpuMetric)
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala:159]: scaladoc changed needed on this one too

**Review by abellina — COMMENTED** [review]
Looks good to me, just nits.

File: docs/tuning-guide.md [hunk 7]
```diff
@@ -133,8 +133,15 @@ Configuration key: [`spark.rapids.sql.concurrentGpuTasks`](configs.md#sql.concur
 Default value: `1`
 
 The RAPIDS Accelerator can further limit the number of tasks that are actively sharing the GPU.
-This is useful for avoiding GPU out of memory errors while still allowing full concurrency for the
-portions of the job that are not executing on the GPU.  Some queries benefit significantly from
+It does this using a semaphore. When metrics or documentation refers to the GPU semaphore it
+is referring to this.
+This restriction is useful for avoiding GPU out of memory errors while still allowing full
+concurrency for the
```
Comment (jlowe) [inline docs/tuning-guide.md:139]: Nit: the wrapping here looks odd if one is examining the .md file directly via a terminal rather than via github's rendering.

**Review by jlowe — COMMENTED** [review]
Overall looks good to me, just a small formatting nit in addition to the other comments.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala (hunk 3 above)
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:374]: I inlined `logical.Range` into it. It is not needed, but made the code a little more robust in my opinion. I can remove it if we want to.

File: docs/tuning-guide.md (hunk 7 above)
Comment (revans2) [reply docs/tuning-guide.md:139]: Feel free to ignore my rant. I will make the change. Really? One of the reasons for mark down is that it hides this when rendered. I have a similar issue just a few lines below. That one is there mostly because I want the diff to be small for reviewing. If we care enough to do it here why not do it everywhere and forget about the diff?

**Comment by revans2** [comment]
build

File: docs/tuning-guide.md (hunk 7 above)
Comment (jlowe) [reply docs/tuning-guide.md]: One of the main advantages of markdown over other doc formats is that is is pretty human-readable in a terminal without requiring rendering in a web-page or other tool, so I'd like to keep it looking nice for those that are looking at things offline.  It is just a nit, and I may be in the minority feeling this way about our markdown formatting.

**Review by jlowe — APPROVED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala (hunk 3 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:374]: no that's fine just curious

**Review by abellina — APPROVED** [review]
